src/synthesizer.ts

```diff
@@ -0,0 +1,406 @@
+import * as path from 'path';
+import * as pulumi from '@pulumi/pulumi';
+import * as cdk from 'aws-cdk-lib/core';
+import { translateCfnTokenToAssetToken } from 'aws-cdk-lib/core/lib/helpers-internal';
+import * as aws from '@pulumi/aws';
+import { CdkConstruct } from './interop';
+import { zipDirectory } from './zip';
+
+/**
+ * Deploy time assets will be put in this S3 bucket prefix
+ * so that separate lifecycle rules can apply
+ */
+const DEPLOY_TIME_PREFIX = 'deploy-time/';
+
+export interface PulumiSynthesizerOptions {
+    /**
+     * A unique identifier for the application that the staging stack belongs to.
+     *
+     * This identifier will be used in the name of staging resources
+     * created for this application, and should be unique across CDK apps.
+     *
+     * The identifier should include lowercase characters, numbers, periods (.) and dashes ('-') only
+     * and have a maximum of 17 characters.
+     */
+    readonly appId: string;
+
+    /**
+     * Explicit name for the staging bucket
+     *
+     * @default - a well-known name unique to this app/env.
+     */
+    readonly stagingBucketName?: string;
+
+    /**
+     * The lifetime for deploy time file assets.
+     *
+     * Assets that are only necessary at deployment time (for instance,
+     * CloudFormation templates and Lambda source code bundles) will be
+     * automatically deleted after this many days. Assets that may be
+     * read from the staging bucket during your application's run time
+     * will not be deleted.
+     *
+     * Set this to the length of time you wish to be able to roll back to
+     * previous versions of your application without having to do a new
+     * `cdk synth` and re-upload of assets.
+     *
+     * @default - Duration.days(30)
+     */
+    readonly deployTimeFileAssetLifetime?: cdk.Duration;
+
+    /**
+     * Specify a custom prefix to be used as the staging stack name and
+     * construct ID. The prefix will be appended before the appId, which
+     * is required to be part of the stack name and construct ID to
+     * ensure uniqueness.
+     *
+     * @default 'staging-stack'
+     */
+    readonly stagingStackNamePrefix?: string;
+
+    /**
+     * Auto deletes objects in the staging S3 bucket and images in the
+     * staging ECR repositories.
+     *
+     * This will also delete the S3 buckets and ECR repositories themselves when
+     * all objects / images are removed.
+     *
+     * @default true
+     */
+    readonly autoDeleteStagingAssets?: boolean;
+}
+
+export class PulumiSynthesizer extends cdk.StackSynthesizer implements cdk.IReusableStackSynthesizer {
```

Comment:
The class deserves a doc comment with a link to the article you attached to this PR, and a brief summary of how it dispatches the synthesizer responsibilities.